test/distributed/org/apache/cassandra/distributed/test/topology/TopologyMixupTest.java

```diff
@@ -0,0 +1,856 @@
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import javax.annotation.Nullable;
+
+import com.google.common.base.Throwables;
+import com.google.common.collect.Sets;
+import org.junit.Test;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import accord.coordinate.Exhausted;
+import accord.local.Node;
+import accord.primitives.Ranges;
+import accord.primitives.TxnId;
+import accord.utils.Gen;
+import accord.utils.Gens;
+import accord.utils.Invariants;
+import accord.utils.Property.Command;
+import accord.utils.Property.Commands;
+import accord.utils.Property.SimpleCommand;
+import accord.utils.RandomSource;
+import org.agrona.collections.Int2ObjectHashMap;
+import org.agrona.collections.IntArrayList;
+import org.agrona.collections.IntHashSet;
+import org.apache.cassandra.config.CassandraRelevantProperties;
+import org.apache.cassandra.config.Config;
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.cql3.ast.Mutation;
+import org.apache.cassandra.cql3.ast.Select;
+import org.apache.cassandra.cql3.ast.Statement;
+import org.apache.cassandra.cql3.ast.Txn;
+import org.apache.cassandra.distributed.Cluster;
+import org.apache.cassandra.distributed.api.ConsistencyLevel;
+import org.apache.cassandra.distributed.api.Feature;
+import org.apache.cassandra.distributed.api.IInstance;
+import org.apache.cassandra.distributed.api.IInvokableInstance;
+import org.apache.cassandra.distributed.api.NodeToolResult;
+import org.apache.cassandra.distributed.api.SimpleQueryResult;
+import org.apache.cassandra.distributed.api.TokenSupplier;
+import org.apache.cassandra.distributed.impl.INodeProvisionStrategy;
+import org.apache.cassandra.distributed.shared.ClusterUtils;
+import org.apache.cassandra.distributed.test.TestBaseImpl;
+import org.apache.cassandra.distributed.test.accord.AccordTestBase;
+import org.apache.cassandra.exceptions.WriteTimeoutException;
+import org.apache.cassandra.io.sstable.format.SSTableFormat;
+import org.apache.cassandra.locator.InetAddressAndPort;
+import org.apache.cassandra.schema.ReplicationParams;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.service.accord.AccordService;
+import org.apache.cassandra.service.accord.api.AccordAgent;
+import org.apache.cassandra.service.consensus.TransactionalMode;
+import org.apache.cassandra.tcm.ClusterMetadata;
+import org.apache.cassandra.tcm.ClusterMetadataService;
+import org.apache.cassandra.utils.CassandraGenerators;
+import org.apache.cassandra.utils.Isolated;
+import org.apache.cassandra.utils.Shared;
+import org.awaitility.Awaitility;
+
+import static accord.utils.Property.multistep;
+import static accord.utils.Property.stateful;
+import static org.apache.cassandra.utils.AccordGenerators.fromQT;
+
+public class TopologyMixupTest extends TestBaseImpl
```

Comment:
I don't think I fully understand this comment, but I am not at all against reusing existing things to improve validation.

> but what I am not seeing is checking if streaming was successful / the data is present on the target nodes. We have a lot of machinery for testing these things in TCM, which can be reused.

If you can point me to stuff, that would be great.  Accord bootstrap is different than normal, but the streaming part is the same... so reusing shouldn't be too hard to get working for accord tables

> and reduce the number of things it does
you mean topology changes, or number of "steps"?

> or refactor it a bit and make it a bit simpler to get overview.

I don't know what you just said.  "get overview"?  what do you mean by this?

> Some middle-ground would be to open a Jira for creating a general framework for driving state changes and integrating it with the rest of our fuzz tooling, that would deduplicate efforts. We have done a lot of work to consolidate things in TCM, so it would be great to reuse some of that.

If there is existing validation logic from TCM to validate that the change was *correct* I am more than happy to reuse.  Each step is single threaded and exists once everything is complete, so wouldn't be hard to add validation checks afterwards to make sure that the cluster is in the right state... I just don't know what logic exists.

Reply:
> If you can point me to stuff, that would be great. Accord bootstrap is different than normal, but the streaming part is the same... so reusing shouldn't be too hard to get working for accord tables

Sure, just take a look at https://github.com/apache/cassandra/blob/trunk/test/distributed/org/apache/cassandra/fuzz/ring/ConsistentBootstrapTest.java#L69-L71

In short, all you need to do is 

```
            ReplayingHistoryBuilder harry = HarryHelper.dataGen(new InJvmSut(cluster),
                                                                new TokenPlacementModel.SimpleReplicationFactor(3),
                                                                SystemUnderTest.ConsistencyLevel.ALL);

                for (int i = 0; i < WRITES; i++)
                    harry.insert();

                harry.validateAll(harry.quiescentLocalChecker());
```

Quiescent local checker (https://github.com/apache/cassandra/blob/42d18a83e4468827da234c68fad3e3058b368d7c/test/harry/main/org/apache/cassandra/harry/sut/injvm/QuiescentLocalStateChecker.java#L37) will infer the placements from the current ring, and will validate that every row is where it should be. In other words, if ownership of a range has changed from node A to node B, it will ensure that the data has moved.

> "get overview"? what do you mean by this?

You are right, I should have expanded on this instead of making a very vague statement. Let me try to remedy this.

First of all, I think the test itself is great, and the only changes I am proposing are optical / structural, and not functional. Just to make it easier for people approaching it for the first time to navigate and expand it.

This class is currently at almost 900LOC. To get an overview is to be able to glance at the class, and understand what it is doing, how to extend it, or what to do when I am trying to debug an issue with the test failure. Maybe I was overly vague, so let me be more specific. I would do the following things: 
  * Most of the methods that create `SimpleCommand` can be made into static factory methods, such as `repairCommand`, `stopInstance`. They can be moved to the top level of the class, and bunched together. This way, anyone looking at the class can see what commands it offers, and navigate to a specific command that interests them quickly.
  * Some code that creates `SimpleCommand` is now inside a `topologyCommand` method. I think that the method itself does make sense, but I also think that individual commands (i.e. `addNode`, `decomission`, `removeNode`, `assasinate`) can benefit from being moved to their own static methods to be near other operations. `topologyCommand` can simply call those methods
  * Expand some of the abbreviations. For example, I'd use `rng` instead of `rs` since it is more commonly used, and expand `th` into `topologyHistory` or just `history`.
  * I'd move enums such as `TopologyChange` and `RemoveType` up, since they help the person understand what's going on in the class, but auxiliary methods such as `asSet`, `executeTxn` and `maybeCreateUDTs`, since they're incidental to the implementation.

With these changes, the person will be able to look at the signatures to better understand what exactly operations are already implemented, and will be able to pick them up and maybe use them in other contexts, for example for testing bootstrap, or decom, etc. Adding a new operation can then also be more straightforward: a person would simply add a new `static` method specifying the operation and then hook it up to the test quickly.

Please let me know if this was helpful!

Reply:
I think I addressed all your comments.  I did the refactors requested and see that the code is easier to read (at least for me).

I also pulled the *accord* bits out and made the class `abstract`.  We now have an accord version and a harry version.

The thinking is that Accord wants to also validate we handle weird and random schemas, so the `ast` generators is better then harry there, but in the "how does topology changes work in Cassndra" version harry works better.

Reply:
@ifesdjeen I think I have address your comment, can you resolve if you agree?  else what is missing?